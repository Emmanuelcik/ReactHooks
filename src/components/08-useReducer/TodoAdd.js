import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: "",
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length <= 1 ){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        }

        handleAddTodo(newTodo);
        reset();
    }
    console.log(description);
    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr/>

            <form onSubmit={handleSubmit}>

                <input
                    className='form-control'
                    type="text"
                    name="description"
                    placeholder="Aprender..."
                    autoComplete='off'
                    onChange={handleInputChange}
                    value={description}
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary d-block mt-2 col-12"
                > 
                    Agregar 
                </button>
            </form>
        </>
    )
}
