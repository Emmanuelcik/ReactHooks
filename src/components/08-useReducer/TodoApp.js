import React, {useEffect, useReducer} from 'react'
import { todoReducer } from './todoReducer';
import "./styles.css";
import { useForm } from '../../hooks/useForm';


const init = () => {

    let dataInLocal = localStorage.getItem("todos");
    return dataInLocal ? JSON.parse(dataInLocal) : [];

    // return [{
    //     id: new Date().getTime(),
    //     desc: "Learning react",
    //     done: false,
    // }];
}

export const TodoApp = () => {

    // const [state, dispatch] = useReducer(reducer, initialState, init)
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    const [ {description}, handleInputChange, reset ] = useForm({
        description: "",
    });

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (todoID) => {
        const action = {
            type: "delete",
            payload: todoID,
        }

        dispatch(action);
    }
    
    console.log(description);

    const handleSubmit = (e) =>  {
        e.preventDefault();

        if(description.trim().length <= 1 ){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        }

        const action = {
            type: "add",
            payload: newTodo
        }

        dispatch(action);
        reset();
    }

    return (
        <div>
            <h1>Todo App ({todos.length})</h1>
            <hr/>
            <div className="row">

                <div className="col col-md-6">
                    <ul className="list-group list-group-flush">
                        {
                            todos.map( (todo, i) => (
                                <li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p className='text-center'>{i + 1}.- { todo.desc} </p>
                                    <button
                                        onClick={() => { handleDelete( todo.id)}}
                                        className="btn btn-danger"
                                    >
                                        Borrar
                                    </button>
                                </li>
                            )) 
                        }
                    </ul>
                </div>
                <div className='col col-md-6'>
                    <h4>Agregar ToDo</h4>
                    <hr/>

                    <form onSubmit={handleSubmit}>

                        <input
                            className='form-control'
                            type="text"
                            name="description"
                            placeholder="Aprender..."
                            autoComplete='off'
                            onChange={handleInputChange}
                            value={description}
                        />

                        <button
                            type="submit"
                            className="btn btn-outline-primary d-block mt-2"
                        > 
                            Agregar 
                        </button>
                    </form>

                </div>

            </div>

        </div>
    )
}
