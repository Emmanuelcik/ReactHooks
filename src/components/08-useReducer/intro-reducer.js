

const initialState = [{
    id: 1,
    todo: "Comprar leche",
    done: false,    
}];

const toDoReducer = (state = initialState, action) => {
    
    if(action?.type === "agregar"){
        return [...state, action.payLoad];
    }

    return state;
}

let todos = toDoReducer();

const newTodo = {
    id: 2,
    todo: "Comprar leche",
    done: false,
}

const action = {
    type: "agregar",
    payLoad: newTodo,
    // newTodo
}

todos = toDoReducer(todos, action);
console.log(todos);
