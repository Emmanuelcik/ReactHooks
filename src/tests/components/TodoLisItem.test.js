import { shallow }  from "enzyme";
import {TodoListItem} from "../../components/08-useReducer/TodoListItem";
import {demoTodos} from "../fixtures/demoTodos";

describe("Testing en TodoListItem", () => {

    const handleToggle = jest.fn();

    const handleDelete = jest.fn();
    
    const wrapper = shallow( <TodoListItem 
            todo={demoTodos[1]} 
            index={1} 
            handleToggle={handleToggle}
            handleDelete={handleDelete}
        />);
    
    test("Should call the handleDelete function", () => {
        
        wrapper.find("button").simulate("click");
        expect(handleDelete ).toHaveBeenCalledWith(demoTodos[1].id);
    });

    test("Should call the handleToggle function", () => {
        wrapper.find("p").simulate("click");
        expect(handleToggle ).toHaveBeenCalledWith(demoTodos[1].id);
    });

})