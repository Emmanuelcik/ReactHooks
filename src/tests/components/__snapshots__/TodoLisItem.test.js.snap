// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing en TodoListItem Should be correctly mounted 1`] = `
<li
  className="list-group-item"
>
  <p
    className=""
    onClick={[Function]}
  >
    2
    .- 
    Aprender mongo
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;

exports[`Testing en TodoListItem Should be correctly mounted 2`] = `
<li
  className="list-group-item"
>
  <p
    className=""
    onClick={[Function]}
  >
    2
    .- 
    Aprender mongo
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
